runtime/include/wasm_memory.h

```diff
@@ -0,0 +1,302 @@
+ * @param offset an offset into the WebAssembly linear memory
+ * @return char at the offset
+ */
+static inline char
+wasm_memory_get_char(struct wasm_memory *self, uint32_t offset)
+{
+	assert(offset + sizeof(char) <= self->size);
+	return *(char *)&self->data[offset];
+}
+
+/**
+ * Get an float from WebAssembly linear memory
+ * @param offset an offset into the WebAssembly linear memory
+ * @return float at the offset
+ */
+static inline float
+wasm_memory_get_f32(struct wasm_memory *self, uint32_t offset)
+{
+	assert(offset + sizeof(float) <= self->size);
+	return *(float *)&self->data[offset];
+}
+
+/**
+ * Get a double from WebAssembly linear memory
+ * @param offset an offset into the WebAssembly linear memory
+ * @return double at the offset
+ */
+static inline double
+wasm_memory_get_f64(struct wasm_memory *self, uint32_t offset)
+{
+	assert(offset + sizeof(double) <= self->size);
+	return *(double *)&self->data[offset];
+}
+
+/**
+ * Get a int8_t from WebAssembly linear memory
+ * @param offset an offset into the WebAssembly linear memory
+ * @return int8_t at the offset
+ */
+static inline int8_t
+wasm_memory_get_i8(struct wasm_memory *self, uint32_t offset)
+{
+	assert(offset + sizeof(int8_t) <= self->size);
+	return *(int8_t *)&self->data[offset];
+}
+
+/**
+ * Get a int16_t from WebAssembly linear memory
+ * @param offset an offset into the WebAssembly linear memory
+ * @return int16_t at the offset
+ */
+static inline int16_t
+wasm_memory_get_i16(struct wasm_memory *self, uint32_t offset)
+{
+	assert(offset + sizeof(int16_t) <= self->size);
+	return *(int16_t *)&self->data[offset];
+}
+
+/**
+ * Get a int32_t from WebAssembly linear memory
+ * @param offset an offset into the WebAssembly linear memory
+ * @return int32_t at the offset
+ */
+static inline int32_t
+wasm_memory_get_i32(struct wasm_memory *self, uint32_t offset)
+{
+	assert(offset + sizeof(int32_t) <= self->size);
+	return *(int32_t *)&self->data[offset];
+}
+
+/**
+ * Get a int32_t from WebAssembly linear memory
+ * @param offset an offset into the WebAssembly linear memory
+ * @return int32_t at the offset
+ */
+static inline int64_t
+wasm_memory_get_i64(struct wasm_memory *self, uint32_t offset)
+{
+	assert(offset + sizeof(int64_t) <= self->size);
+	return *(int64_t *)&self->data[offset];
+}
+
+static inline uint32_t
+wasm_memory_get_page_count(struct wasm_memory *self)
+{
+	return (uint32_t)(self->size / WASM_PAGE_SIZE);
+}
+
+/**
+ * Get a null-terminated String from WebAssembly linear memory
+ * @param offset an offset into the WebAssembly linear memory
+ * @param size the maximum expected length in characters
+ * @return pointer to the string or NULL if max_length is reached without finding null-terminator
+ */
+static inline char *
+wasm_memory_get_string(struct wasm_memory *self, uint32_t offset, uint32_t size)
+{
+	assert(offset + (sizeof(char) * size) <= self->size);
+
+	for (uint32_t i = 0; i < size; i++) {
```

Comment:
isn't this effectively `strnlen`? glibc will have very fast versions by default.